src/xml/writer.cr

```diff
@@ -0,0 +1,312 @@
+    # invokes the block and then emits the end of the element.
+    def element(__prefix__ : String?, __name__ : String, __namespace_uri__ : String?, **attributes)
+      element(__prefix__, __name__, __namespace_uri__, attributes) do
+        yield
+      end
+    end
+
+    # ditto
+    def element(__prefix__ : String?, __name__ : String, __namespace_uri__ : String?, attributes : Hash | NamedTuple)
+      start_element __prefix__, __name__, __namespace_uri__
+      attributes(attributes)
+      yield.tap { end_element }
+    end
+
+    # Emits an element with namespace info with the given attributes.
+    def element(prefix : String?, name : String, namespace_uri : String?, **attributes)
+      element(prefix, name, namespace_uri, attributes)
+    end
+
+    # ditto
+    def element(prefix : String?, name : String, namespace_uri : String?, attributes : Hash | NamedTuple)
+      start_element(prefix, name, namespace_uri)
+      attributes(attributes)
+      end_element
+    end
+
+    # Emits the start attribute of an attribute.
+    def start_attribute(name : String) : Nil
+      call StartAttribute, name
+    end
+
+    # Emits the start attribute of an attribute with namespace info.
+    def start_attribute(prefix : String?, name : String, namesapce_uri : String?)
+      call StartAttributeNS, string_to_unsafe(prefix), name, string_to_unsafe(namesapce_uri)
+    end
+
+    # Emits the end of an attribute.
+    def end_attribute : Nil
+      call EndAttribute
+    end
+
+    # Emits the start of an attribute, invokes the block,
+    # and then emits the end of the attribute.
+    def attribute(*args, **nargs)
+      start_attribute *args, **nargs
+      yield.tap { end_attribute }
+    end
+
+    # Emits an attribute with a value.
+    def attribute(name : String, value) : Nil
+      call WriteAttribute, name, value.to_s
+    end
+
+    # Emits an attribute with namespace info and a value.
+    def attribute(prefix : String?, name : String, namesapce_uri : String?, value) : Nil
+      call WriteAttributeNS, string_to_unsafe(prefix), name, string_to_unsafe(namesapce_uri), value.to_s
+    end
+
+    # Emits the given attributes with their values.
+    def attributes(**attributes)
+      attributes(attributes)
+    end
+
+    # ditto
+    def attributes(attributes : Hash | NamedTuple)
+      attributes.each do |key, value|
+        attribute key.to_s, value
+      end
+    end
+
+    # Emits text content.
+    #
+    # Text content can happen inside of an element, attribute value, cdata, dtd, etc.
+    def text(content : String) : Nil
+      call WriteString, content
+    end
+
+    # Emits the start of a CDATA section.
+    def start_cdata : Nil
+      call StartCDATA
+    end
+
+    # Emits the end of a CDATA section.
+    def end_cdata : Nil
+      call EndCDATA
+    end
+
+    # Emits the start of a CDATA section, invokes the block
+    # and then emits the enf of the CDATA section.
```

Comment:
"emits the end" or maybe "closes the CDATA section"? Same for other occurrences.